src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionPool.cs

```diff
@@ -370,10 +381,45 @@ public byte[] Http2AltSvcOriginUri
             }
         }
 
-        public bool EnableMultipleHttp2Connections => _poolManager.Settings.EnableMultipleHttp2Connections;
+        private bool EnableMultipleHttp2Connections => _poolManager.Settings.EnableMultipleHttp2Connections;
 
         /// <summary>Object used to synchronize access to state in the pool.</summary>
-        private object SyncObj => _idleConnections;
+        private object SyncObj => _availableHttp11Connections;
+
+        // Overview of connection management (mostly HTTP version independent):
+        //
+        // Each version of HTTP (1.1, 2, 3) has its own connection pool, and each of these work in a similar manner,
+        // allowing for differences between the versions (most notably, HTTP/1.1 is not multiplexed.)
+        //
+        // When a request is submitted for a particular version (e.g. HTTP/1.1), we first look in the pool for available connections.
+        // An "available" connection is one that is (hopefully) usable for a new request.
+        //      For HTTP/1.1, this is just an idle connection.
+        //      For HTTP2/3, this is a connection that (hopefully) has available streams to use for new requests.
+        // If we find an available connection, we will attempt to validate it and then use it.
+        //      We check the lifetime of the connection and discard it if the lifetime is exceeded.
+        //      We check that the connection has not shut down; if so we discard it.
+        //      For HTTP2/3, we reserve a stream on the connection. If this fails, we cannot use the connection right now.
+        // If validation fails, we will attempt to find a different available connection.
+        //
+        // Once we have found a usable connection, we use it to process the request.
+        //      For HTTP/1.1, a connection can handle only a single request at a time, thus it is immediately removed from the list of available connections.
+        //      For HTTP2/3, a connection is only removed from the available list when it has no more available streams.
```

Comment:
What happens if the server dynamically changes how many streams are allowed on the connection?

Reply:
See Http2Connection.SendHeadersAsync. We will throw a retryable exception.

Reply:
Also note this can't happen in HTTP3.